exercises/IaC/ansible/variables/.templates/README.md.erb

```diff
@@ -0,0 +1,378 @@
+
+The playbook in the previous example, whilst flexible, contained a code smell in that the Inventory file duplicates the `app_env` variable for each host.
+
+<%
+revised_inventory=<<YAML
+[dev-servers]
+#{dev_server}
+
+[prod-servers]
+prod1
+prod2
+YAML
+%>
+Both the Inventory and Playbook can be dried up by making use of Ansible's `group_vars` feature. Update `<%= write_to_file('ansible/inventory', revised_inventory) %>` to remove the declaration of `app_env` for each host. This should mean that the inventory now looks like the following:
+```YAML
+<%= revised_inventory %>
+```
+
+The following group variables files can be used instead.
+
+<%
+prod_group_vars=<<YAML
+prompt: PROD
+YAML
+
+dev_group_vars=<<YAML
+prompt: DEV
+YAML
+%>
+
+<%= write_to_file('ansible/group_vars/prod-servers.yml', prod_group_vars) %>
+```YAML
+<%= prod_group_vars %>
+```
+<%= write_to_file('ansible/group_vars/dev-servers.yml', dev_group_vars) %>
+```YAML
+<%= dev_group_vars %>
+```
+
+<%
+  revised_playbook=<<YAML
+---
+- hosts: all
+  tasks:
+    - name: Set root prompt
+      lineinfile:
+        dest: /root/.bashrc
+        line: 'PS1="[{{prompt}}]$ "'
+        insertbefore: EOF
+
+YAML
+%>
+
+Now that `app_env` is no longer required we can also remove the `include_vars` statement from the playbook. Update <%=write_to_file('ansible/prompt.yml', revised_playbook) %> to read as follows:
+
+```YAML
+<%= revised_playbook %>
+```
+
+Run `<%= command(cic_down_command) %>` and then `<%= command(cic_up_command) %>` to reset the test infrastructure and rerun <%= command('ansible-playbook ansible/prompt.yml -i ansible/inventory') %>
+```
+<%= last_command_output %>
+```
+
+Use the `<%=cic_connect_command %>` command to connect to any of the servers and you will see that their prompts continue to be set correctly.
+
+## Now it's your turn!
+The operations team are looking after the development and production environments for an application. The two environments are made up 2 development server and 2 production servers.
+
+Their hostnames are:
+
+- dev-app1
+- dev-app2
+- prod-app1
+- prod-app2
+
+The application deployed to these environments requires a file to be located at `/etc/app.id`. This file should containing a single line in the format:
+
+```APPSTRING:ENVIRONMENT:HOSTNAME```
+
+- `APPSTRING` is a static string - almost always `ABCAPP`, but sometimes the team need to change this for ad-hoc testing, so they would like this to default to ```CIC_WEBAPP``` but be overridable from the command line.
+
+- `ENVIRONMENT` is the environment that the server is a member of. This should be DEV for development machines and PROD for production machines.
```

Comment:
_environment_ appears to be a special word in ansible and is not resolved as a normal variable - It is worth having a note to this effect, or perhaps using app_env instead of _environment_ to that people don't go down the wrong path by default.

Reply:
Should probably call out to the doc about restricted keywords here as _environment_ is not the only one. https://docs.ansible.com/ansible/2.5/reference_appendices/playbooks_keywords.html

Reply:
I have put in a note warning participants against choosing a variable name that clashes with those used by ansible:

> **Note:** Be careful not to choose a variable name that is one of [the variable names used by Ansible](https://docs.ansible.com/ansible/2.5/reference_appendices/playbooks_keywords.html) as doing so can lead to unexpected results.